flink/src/main/java/org/apache/iceberg/flink/connector/sink/IcebergWriter.java

```diff
@@ -0,0 +1,330 @@
+        namespace, tableName, subtaskId, openPartitionFiles.size());
+    // close all open files without sending DataFile list to downstream committer operator.
+    // because there are not checkpointed,
+    // we don't want to commit these files.
+    for (Map.Entry<String, FileWriter> entry : openPartitionFiles.entrySet()) {
+      final FileWriter writer = entry.getValue();
+      final Path path = writer.getPath();
+      try {
+        LOG.debug("Iceberg writer {}.{} subtask {} start to abort file: {}",
+            namespace, tableName, subtaskId, path);
+        writer.abort();
+        LOG.info("Iceberg writer {}.{} subtask {} completed aborting file: {}",
+            namespace, tableName, subtaskId, path);
+      } catch (Throwable t) {
+//                LOG.error(String.format("Iceberg writer %s.%s subtask %d failed to abort open file: %s",
+//                        namespace, tableName, subtaskId, path.toString()), t);
+        LOG.error("Iceberg writer {}.{} subtask {} failed to abort open file: {}. Throwable = {}",
+            namespace, tableName, subtaskId, path.toString(), t);
+        continue;
+      }
+
+      try {
+        LOG.debug("Iceberg writer {}.{} subtask {} deleting aborted file: {}",
+            namespace, tableName, subtaskId, path);
+        io.deleteFile(path.toString());
+        LOG.info("Iceberg writer {}.{} subtask {} deleted aborted file: {}",
+            namespace, tableName, subtaskId, path);
+      } catch (Throwable t) {
+//                LOG.error(String.format(
+//                        "Iceberg writer %s.%s subtask %d failed to delete aborted file: %s",
+//                        namespace, tableName, subtaskId, path.toString()), t);
+        LOG.error("Iceberg writer {}.{} subtask {} failed to delete aborted file: {}. Throwable = {}",
+            namespace, tableName, subtaskId, path.toString(), t);
+      }
+    }
+    LOG.info("Iceberg writer {}.{} subtask {} aborted {} open files",
+        namespace, tableName, subtaskId, openPartitionFiles.size());
+    openPartitionFiles.clear();
+  }
+
+  @Override
+  public void processElement(StreamRecord<IN> element) throws Exception {
+    IN value = element.getValue();
+    try {
+      processInternal(value);
+    } catch (Exception t) {
+      if (!skipIncompatibleRecord) {
+        throw t;
+      }
+    }
+  }
+
+  @VisibleForTesting
+  void processInternal(IN value) throws Exception {
+    Record record = serializer.serialize(value, schema);
+    processRecord(record);
+  }
+
+  /**
+   * process element as {@link Record}
+   */
+  private void processRecord(Record record) throws Exception {
+    if (partitioner == null) {
+      partitioner = new RecordPartitioner(spec);
+    }
+    partitioner.partition(record);
+    final String partitionPath = locations.newDataLocation(spec, partitioner, "");
+    if (!openPartitionFiles.containsKey(partitionPath)) {
+      final Path path = new Path(partitionPath, generateFileName());
```

Comment:
I wonder whether we have a max number of `openPartitionFiles` here, because this may introduce performance regression when too many files are procesed. (random reads/writes)

Reply:
there is no max limit. here the number of open files is largely determined by table partitioning spec. there are a few examples
* partition by event time and there is long tail problem for event time
* partition by some high cardinality columns, like country.

Main concern for large number of open files (per subtask) is the memory footprint, especially when writing as Parquet format. There is also concern of many small files.